<a href="https://colab.research.google.com/github/chaitanysaraswat/Advanced-Analytics-Using-Statistics/blob/main/lecture11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1-sample z test/ 1 sample t test
This test is used to compare the sample mean of one sample sample against a calim value and establish whether their difference is statistically significant or not

2-sample t test
This test is used to compare the difference of means of two independent samples against a claim value and establish whether their difference is statistically significant or not

Paired t test
This test is used when we have two paired samples. The average difference between the paired values is compared against a claim values and establish whether their difference is statistically significant or not
The key word here is before-after

One way ANOVA
Used when we have multiple samples. This test compares all the pair of the means. It checks whether the difference of each pair of mean can be ignored or not OR can be considered to be equal to 0 or not.
The Ho in this case is that all the pair of means are equal

Multi way ANOVA
Used when we have multiple factors influencing the response. IN addition to their individual effect, we would also like to study the interactions between various factors.

In [ ]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None)
from numpy import random
from collections import Counter
from numpy.linalg import inv
from numpy.linalg import eig
import matplotlib
from matplotlib import pyplot as plt
import seaborn as sns
import pylab
from pylab import legend
from pylab import plot, show, title, xlabel, ylabel
import scipy
from scipy import stats
from scipy.stats import binom,poisson,norm,t,expon,f
from sklearn.model_selection import train_test_split
import statsmodels.api as sample_data
from statsmodels.formula.api import ols
import statsmodels.stats.multicomp
from statsmodels.stats.weightstats import ttost_paired
import statsmodels.stats.weightstats as ssw
from statsmodels.stats.multicomp import pairwise_tukeyhsd
import statsmodels.stats
from statsmodels.stats.proportion import proportions_ztest

Inferential:


paired t test:
- this test is used when we have two paired samples avg diff b/w paired values
  is compared against a claim value to establish whether the diff statstically significant or not.
 - the keyword here is "before-after".
 - sample size needs to be size  source of data needs to be same

Time taken by individual will be coming by atleast by  2min

h0(null hypothesis) before -after >=2min

Q procesing will be reduced by atleast 1min. h0: before-after >=1min right

Q performance has reduced by  not more than 1min. h0: before-after<=1min left tail test

after-before>=10


In [ ]:
before=np.array([12,14,10,9,8,13])
after=np.array([11.8,9.7,11.3,8.2,7.2,12.3])


In [ ]:
d=(before-after)
np.mean(d)

np.float64(0.9166666666666665)

h0  avg(before-after)>2

now we wil check whether the 0.911 and >2 above is matter of chance variation or not

In [ ]:

ssw.ttost_paired(before, after, low=-2, upp=2)

(np.float64(0.10453031697930062),
 (np.float64(3.8804663986357473),
  np.float64(0.005818213196929444),
  np.float64(5.0)),
 (np.float64(-1.4413160909218492),
  np.float64(0.10453031697930062),
  np.float64(5.0)))

Pvalue>0.05 so we do not reject null hypothesis

df is degree of freedom which is (no of values-1)

In [ ]:
scipy.stats.ttest_1samp(before-after,popmean=2,alternative="less")

TtestResult(statistic=np.float64(-1.4413160909218492), pvalue=np.float64(0.10453031697930062), df=np.int64(5))

In [ ]:
(np.mean(d)-2)/((np.std(d,ddof=1))/6**0.5)
# (mean(diff)-claim_mean)/std(d)/number**0.5

np.float64(-1.4413160909218492)

# ANOVA ( ANalysis Of Variance )

pupose:
- this test is used when we have multiple samples.
- this test will compare all the pair of the means and check wheather they can be considered equal or not.
- standard null hpothesis all the pair means are equal

In [ ]:
dfA = pd.DataFrame({'Col1':np.repeat('A',7),'Col2':np.random.randint(10,30,7)}) # col1 will have a is 7 no of times A and col 2 will 7 random value
dfB = pd.DataFrame({'Col1':np.repeat('B',9),'Col2':np.random.randint(20,40,9)}) # col1 will have a is 9 no of times A and col 2 will 9 random value
dfC = pd.DataFrame({'Col1':np.repeat('C',8),'Col2':np.random.randint(15,30,8)}) # col1 will have a is 8 no of times A and col 2 will 8 random value
df = pd.concat([dfA,dfB,dfC])


In [ ]:
df

,Col1,Col2
0,A,15
1,A,24
2,A,22
3,A,19
4,A,27
5,A,20
6,A,17
0,B,39
1,B,39
2,B,21


h0 mean(a)=mean(b)

Anova below method
- Test Statstics will be f below method
- f=variations b/w samples/variations within sample
- deviation of each datapoint from its own sample mean is variation within sample.it is denoted by w in pic
-  variation each sample mean from the grand mean which is denoted by b in pic


- as the value of test statstics increases it indicates that we are going away from epected result so p value will decrease
case1: if between variation of samples is more and within variation is less in sample TTHEN VALUE OF F WILL INCREASE



# -** f distribution is always ratio of two variances**


In [ ]:
import statsmodels.api as sm
mod1 = ols('Col2 ~ Col1',df).fit()
tbl = sm.stats.anova_lm(mod1)
print(tbl)

PatsyError: Error evaluating factor: NameError: name 'Col1' is not defined
    Col2 ~ Col1
           ^^^^

In [ ]:
print(pairwise_tukeyhsd(df.Col2, df.Col1))

 Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower    upper  reject
-----------------------------------------------------
     A      B    8.873 0.0167   1.5004 16.2456   True
     A      C   0.6786 0.9723   -6.893  8.2501  False
     B      C  -8.1944 0.0221 -15.3032 -1.0857   True
-----------------------------------------------------


ANOVA using excel sheet

-

In [ ]:
df1=pd.read_excel('/content/drive/MyDrive/My pitch/data_anova (1).xlsx')

In [ ]:
df1.head(20)

,Col1,Col2
0,A,10
1,A,23
2,A,14
3,A,25
4,A,22
5,A,11
6,A,12
7,B,26
8,B,39
9,B,29


In [ ]:
df1.index=np.arange(df.shape[0])

source->db->etl->t->spark,pythoncollab,grafana,kibana,prometheus,powerbi,tableau

In [ ]:
grp1=df1.groupby("Col1")

In [ ]:
sm=grp1.agg('mean')
sm

,Col2
Col1,
A,16.714286
B,31.222222
C,23.250000


In [ ]:
gm=np.mean(df1.Col2)
gm

np.float64(24.333333333333332)

In [ ]:
sm.loc['A']

,A
Col2,16.714286


In [ ]:
7*(sm.loc['A'] - gm)**2 + 9*(sm.loc['B'] - gm)**2 + 8*(sm.loc['C'] - gm)**2

,0
Col2,842.849206


In [ ]:
sampA = df1.Col2[np.where(df.Col1=='A')[0]]
sampB = df1.Col2[np.where(df.Col1=='B')[0]]
sampC = df1.Col2[np.where(df.Col1=='C')[0]]

In [ ]:
sum((np.array(sampA) - np.array(sm.loc['A']))**2) + sum((np.array(sampB) - np.array(sm.loc['B']))**2) + sum((np.array(sampC) - np.array(sm.loc['C']))**2)


ANOVA TABLE DESCRIPTION:

COL1 SUM SQ COLUM=>842.849(B/W SAMPLES)

RESIDUAL SUMSQ COLUMN =>684 (WITHIN SAMPLES)

COL1

mean squared=sample stats/degree of freedom


In [ ]:
import statsmodels.api as sm
from statsmodels.formula.api import ols
mod1=ols('Col2 ~ Col1',df).fit()
tbl=sm.stats.anova_lm(mod1)
tbl

,df,sum_sq,mean_sq,F,PR(>F)
Col1,2.0,409.188492,204.594246,6.073307,0.008293
Residual,21.0,707.436508,33.687453,NaN,NaN


# one way anova

is used when the response is continuos data
sampling will done on categorical data

Assumption one anova :

- data for all the sample need to be normally distributed
- variance of all the samples need to be equal

# multiple anova:

whether gender seffects salary
whether

In [ ]:
zoo = [34,43,57,40,85,68,67,53,41,24,42,52]
supp=[1,1,1,1,2,2,2,2,3,3,3,3]
lake=["R","D","R","D","R","D","R","D","R","D","R","D"]



In [ ]:
df.head()

,Col1,Col2
0,A,15
1,A,24
2,A,22
3,A,19
4,A,27


In [ ]:
df.nunique()

,0
Col1,3
Col2,15


In [ ]:
df=pd.DataFrame({'Resp':zoo,"Factor1":supp,"Factor2":lake})
model = ols('Resp~Factor1*Factor2',df).fit()
res=sm.stats.anova_lm(model)

when the change of level of one factor  has different impact on levels of other factor then we say two factors are interacting

change of the level factor from dry to rain  has different impact on the car and bike which are the levels of factor mode of transport so we can say mode and whether interact with eachother.



## interview question:
Q why can't we use 2 sample test for 3 samples rather using one way anova?

becoz error percentage will again and again 2 sample t test eroor will get inflated in on anova error is 0.05



interaction:

response are influence  by factor or not and vice versa

## one sample variance test pending
## two sample variance test

- for binomial distribution: mean=n*p std= sqrt(n*p*i-p)

## Discrete data validation:

*   one proporation t test :-   fraction of 2 discrete data

this test is used to compare a proportion calculated from sample  against a claim value and establish whether the diff is statistically significant or not.
- normal distribution is used when we have to use approx in p-value
*   two proportion t test :-   



In [ ]:
mu=500*0.7
sd=(500*0.7*0.3)**0.5


In [ ]:
binom.cdf(370,500,0.7)-binom.cdf(329,500,0.7)

np.float64(0.9547133734363867)

In [ ]:
norm.cdf(370.5,mu,sd)-norm.cdf(329.5,mu,sd)

np.float64(0.9545639632657537)

40% students coming from maharashtra

null hypothesis => Ho <=0.4 ha>=0.4

in a saple of 400 student 180 are from maharashtra

In [ ]:
proportion= 180/400
proportion

0.45

In [ ]:
from statsmodels.stats.proportion import proportions_ztest
proportions_ztest(180, 400, 0.4, alternative='larger', prop_var=0.4)

(np.float64(2.0412414523193148), np.float64(0.02061341666858185))

In [ ]:
1-norm.cdf(2.0412414523193148) #will return area in which where the value  it can be found or "p -value"

np.float64(0.020613416668581852)

## 2-proportion test
This test is used to compare the difference of two proportions calculated from two samples against a claim value and establish whether their difference is statistically significant or not

Q pune CDAC claims taht they have atleast 8%more engineers as compared to cdac kharghar

- ho pr(pune)-pr(kharghar)>=0.8
- ha < 0.08

in a sample of 80 student in kharghar ,48 were btech

in a sample of 70 student in pune ,40 were btech

In [ ]:
40/70-48/80

-0.02857142857142858

In [ ]:
engg = np.array([40,48])
total = np.array([70,80])
statsmodels.stats.proportion.proportions_ztest(engg,total,0.08,alternative='smaller')

(np.float64(-1.3471563869719667), np.float64(0.08896493545619544))

Chi square test for goodness of fit:

Used when there are multiple categories. The counts obtained for each category would be compared aginst the claimed pattern to check whether the counts follow the pattern or not. The standard Ho is that the counts follow the pattern

claimed pattern for the qualifictions

btech -0.4 mtech-0.1 mca-0.3 msc-0.2

btech -50 mtech-20 mca-45 msc-25

whether counts obtained from sample follow the pattern or not

In [ ]:
counts=np.array([50,20,45,25])
pattern=np.array([0.4,0.1,0.3,0.2])
exp_counts=sum(counts)*pattern
exp_counts

array([56., 14., 42., 28.])

In [ ]:
from scipy.stats import chisquare
scipy.stats.chisquare(counts,f_exp=exp_counts)

In [ ]:
counts=np.array([70,10,35,45])
exp_counts=sum(counts)*pattern
exp_counts

array([64., 16., 48., 32.])

In [ ]:
from scipy.stats import chisquare
scipy.stats.chisquare(counts,f_exp=exp_counts)

Power_divergenceResult(statistic=np.float64(11.614583333333334), pvalue=np.float64(0.008827096989130126))

In [ ]:
sum((exp_counts - counts)**2/exp_counts)
(exp_counts - counts)**2/exp_counts

array([0.5625    , 2.25      , 3.52083333, 5.28125   ])

## chi square Degree of associciations:
- eg gender anad effect on food u eat
- there will be 2 variable
- null hypothesis categorical data a that there is
- Chi square test for degree pf association
Used when we wish to check and establish whether one categorical data has an effect on the other categorical data or not

In [ ]:
from scipy.stats import chi2_contingency
food=np.array([[45,70,60],[70,40,25]])
food

array([[45, 70, 60],
       [70, 40, 25]])

degree of freedom = (row-1)*(columns-1)

In [ ]:
scipy.stats.chi2_contingency(food)

Chi2ContingencyResult(statistic=np.float64(23.25424258416586), pvalue=np.float64(8.920832446902749e-06), dof=2, expected_freq=array([[64.91935484, 62.09677419, 47.98387097],
       [50.08064516, 47.90322581, 37.01612903]]))